routing/state.go

```diff
@@ -0,0 +1,157 @@
+package routing
+
+import (
+	"errors"
+	"github.com/gogo/protobuf/proto"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	pb "github.com/libp2p/go-libp2p-core/routing/pb"
+	ma "github.com/multiformats/go-multiaddr"
+	"time"
+)
+
+// The domain string used for routing state records contained in a SignedEnvelope.
+const StateEnvelopeDomain = "libp2p-routing-state"
+
+// The type hint used to identify routing state records in a SignedEnvelope.
+// TODO: register multicodec
+var StateEnvelopePayloadType = []byte("/libp2p/routing-state-record")
+
+type SignedRoutingState struct {
+	// PeerID is the ID of the peer this record pertains to.
+	PeerID peer.ID
+
+	// Seq is an increment-only sequence counter used to order RoutingState records in time.
+	Seq uint64
+
+	// Addrs contains the public addresses of the peer this record pertains to.
+	Addrs []ma.Multiaddr
+
+	// Envelope contains the signature and serialized RoutingStateRecord protobuf.
+	// Although it uses a bit
+	Envelope *crypto.SignedEnvelope
```

Comment:
This is weird. The envelope should wrap the payload, not the other way around.

Reply:
Keeping the original envelope around lets us marshal / unmarshal it without invalidating the signature.  Essentially `SignedRoutingState` is a composition of a `SignedEnvelope` with an unpacked representation of the payload.

I agree this is weird though, and it looks like I got distracted and didn't finish writing the comment to explain it 😄 .

I'm pretty sure the reason I ended up with this was to enable an api like:

```go
type CertifiedAddrBook interface {
  AddCertifiedAddrs(record *SignedRoutingState)
}
```

instead of `AddCertifiedAddrs(envelope *SignedEnvelope)`, where `AddCertifiedAddrs` would be responsible for "opening" the envelope. 

Although looking at it now, that might actually be easier to use, since the caller of `AddCertifiedAddrs` wouldn't have to open the envelope first...